indexing-service/src/main/java/io/druid/indexing/overlord/http/OverlordResource.java

```diff
@@ -575,58 +537,215 @@ public Response getCompleteTasks(
+        duration = theInterval.toDuration();
+      }
+      final List<TaskInfo<Task>> taskInfoList = taskStorageQueryAdapter.getRecentlyCompletedTaskInfo(
+          maxCompletedTasks, duration);
+      for (TaskInfo taskInfo : taskInfoList) {
+        taskInfoCache.putIfAbsent(taskInfo.getId(), taskInfo.getDataSource());
+      }
+      final List<TaskStatusPlus> completedTasks = Lists.transform(taskInfoList, completeTaskTransformFunc);
+      finalTaskList.addAll(completedTasks);
+    }
 
-    final List<TaskStatusPlus> completeTasks = Lists.newArrayList(Iterables.transform(
-        recentlyFinishedTasks,
-        status -> {
-          final Pair<DateTime, String> pair = taskStorageQueryAdapter.getCreatedDateAndDataSource(status.getId());
-          return new TaskStatusPlus(
-              status.getId(),
-              taskFunction.apply(status.getId()).getType(),
-              pair.lhs,
-              // Would be nice to include the real queue insertion time, but the
-              // TaskStorage API doesn't yet allow it.
+    List<TaskInfo<Task>> allActiveTaskInfo = new ArrayList<>();
+    if (state == null || !"complete".equals(StringUtils.toLowerCase(state))) {
+      allActiveTaskInfo = taskStorageQueryAdapter.getActiveTaskInfo();
+      for (TaskInfo taskInfo : allActiveTaskInfo) {
+        taskInfoCache.putIfAbsent(taskInfo.getId(), taskInfo.getDataSource());
+      }
+    }
+    if (state == null || "waiting".equals(StringUtils.toLowerCase(state))) {
+      final List<AnyTask> waitingWorkItems = filterActiveTasks(req, RunnerTaskState.WAITING, allActiveTaskInfo);
+      List<TaskStatusPlus> transformedWaitingList = Lists.transform(waitingWorkItems, activeTaskTransformFunc);
+      finalTaskList.addAll(transformedWaitingList);
+    }
+    if (state == null || "pending".equals(StringUtils.toLowerCase(state))) {
+      final List<AnyTask> pendingWorkItems = filterActiveTasks(req, RunnerTaskState.PENDING, allActiveTaskInfo);
+      List<TaskStatusPlus> transformedPendingList = Lists.transform(pendingWorkItems, activeTaskTransformFunc);
+      finalTaskList.addAll(transformedPendingList);
+    }
+    if (state == null || "running".equals(StringUtils.toLowerCase(state))) {
+      final List<AnyTask> runningWorkItems = filterActiveTasks(req, RunnerTaskState.RUNNING, allActiveTaskInfo);
+      List<TaskStatusPlus> transformedRunningList = Lists.transform(runningWorkItems, activeTaskTransformFunc);
+      finalTaskList.addAll(transformedRunningList);
+    }
+    final List<TaskStatusPlus> authorizedList = securedTaskStatusPlus(
+        finalTaskList,
+        taskInfoCache,
+        dataSource,
+        type,
+        req
+    );
+    return Response.ok(authorizedList).build();
+  }
+
+  private List<AnyTask> filterActiveTasks(
+      HttpServletRequest req,
+      RunnerTaskState state,
+      List<TaskInfo<Task>> allActiveTaskInfo
+  )
+  {
+    final List<AnyTask> allTasks = Lists.newArrayList();
```

Comment:
Let's move `allTasks` and the for loop below to `getTasks`, so we only create this list once

Reply:
good point, moved inside `getTasks`